flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/utils/ExpressionTestBase.scala

```diff
@@ -108,95 +108,100 @@ abstract class ExpressionTestBase {
+        classOf[MapFunction[RowData, BinaryRowData]],
+        bodyCode,
+        resultType,
+        inputType)
+
+      val mapper = genFunc.newInstance(getClass.getClassLoader)
+
+      val isRichFunction = mapper.isInstanceOf[RichFunction]
+
+      // call setRuntimeContext method and open method for RichFunction
+      if (isRichFunction) {
+        val richMapper = mapper.asInstanceOf[RichMapFunction[_, _]]
+        val t = new RuntimeUDFContext(
+          new TaskInfo("ExpressionTest", 1, 0, 1, 1),
+          classOf[ExpressionTestBase].getClassLoader,
+          env.getConfig,
+          Collections.emptyMap(),
+          Collections.emptyMap(),
+          null)
+        richMapper.setRuntimeContext(t)
+        richMapper.open(new Configuration())
+      }
 
-    val result = mapper.map(testRow)
+      val testRow = if (containsLegacyTypes) {
+        val converter = DataFormatConverters
+          .getConverterForDataType(resolvedDataType)
+          .asInstanceOf[DataFormatConverter[RowData, Row]]
+        converter.toInternal(testData)
+      } else {
+        val converter = DataStructureConverters
+          .getConverter(resolvedDataType)
+          .asInstanceOf[DataStructureConverter[RowData, Row]]
+        converter.toInternalOrNull(testData)
+      }
 
-    // call close method for RichFunction
-    if (isRichFunction) {
-      mapper.asInstanceOf[RichMapFunction[_, _]].close()
-    }
+      val result = mapper.map(testRow)
 
-    // compare
-    testExprs
-      .zipWithIndex
-      .foreach {
-        case ((originalExpr, optimizedExpr, expected), index) =>
-
-          // adapt string result
-          val actual = if(!result.asInstanceOf[BinaryRowData].isNullAt(index)) {
-            result.asInstanceOf[BinaryRowData].getString(index).toString
-          } else {
-            null
-          }
-
-          val original = if (originalExpr == null) "" else s"for: [$originalExpr]"
-
-          assertEquals(
-            s"Wrong result $original optimized to: [$optimizedExpr]",
-            expected,
-            if (actual == null) "null" else actual)
+      // call close method for RichFunction
+      if (isRichFunction) {
+        mapper.asInstanceOf[RichMapFunction[_, _]].close()
       }
+
+      // compare
+      testExprs
+        .zipWithIndex
+        .foreach {
+          case ((originalExpr, optimizedExpr, expected), index) =>
+
+            // adapt string result
+            val actual = if(!result.asInstanceOf[BinaryRowData].isNullAt(index)) {
+              result.asInstanceOf[BinaryRowData].getString(index).toString
+            } else {
+              null
+            }
+
+            val original = if (originalExpr == null) "" else s"for: [$originalExpr]"
+
+            assertEquals(
+              s"Wrong result $original optimized to: [$optimizedExpr]",
+              expected,
+              if (actual == null) "null" else actual)
+        }
+    } finally {
+      testExprs.clear()
```

Comment:
Is this necessary? I noticed we will clear this in `@Before`

Reply:
Yes, currently  the test suite is collecting all expressions firstly and then evaluate once, if we want to check exception for   multiple expressions in one unit test, the test suite always  cover the first expression.

Reply:
I think we need to find another way to test exception. 
`testExpectExceptionThrown` is breaking the testing loop of this test base, which is triggered by `@Before` and `@After`. If we write a test case mixing with `testExpectExceptionThrown` and some normal tests, I suspect there will be some weird stuff. 
One choice is you collect all tests which expected with exception into another array other than `testExprs`, and also verify this exception array in `@After`.